clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -3794,8 +3793,11 @@ void handleResponse(AbstractResponse abstractResponse) {
 
                         KafkaFutureImpl<ListOffsetsResultInfo> future = futures.get(tp);
                         Errors error = partitionData.error;
-                        if (MetadataOperationContext.shouldRefreshMetadata(error) && topicPartitionOffsets.get(tp) != null) {
-                            partitionsWithErrors.add(tp);
+                        if (topicPartitionOffsets.get(tp) == null) {
```

Comment:
Would this be any clearer?

```java
OffsetSpec offsetRequestSpec = topicPartitionOffsets.get(tp);
if (offsetRequestSpec == null) {
  future.completeExceptionally(error.exception());
} else if (shouldRefreshMetadata(error) {
  retryTopicPartitionOffsets.put(tp, offsetRequestSpec);
} else {
...
```

Also, in the case of that we got back an unexpected offset, I think we can raise a new `KafkaException` and provide a clear message indicating what happened.

Reply:
Yeah that's probably a better way of writing this. When you say raising a new `KafkaException`, do you mean throw it within the call or calling `fail` that raises a `KafkaException` here: https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java#L707)

Reply:
Currently in this case, we do the following:
```
future.completeExceptionally(error.exception());
```
I am suggesting we do something like this:
```
future.completeExceptionally(new KafkaException("Unexpected partition in response....");
```

Reply:
Makes sense, just fixed that.